### PR TITLE
[core] Introduce bucket level value filter (filter push down) in primary key table

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose
Introduce bucket level value filter in primary key.

Current:  primary key does not support value filter.

Improve: if the bucket range min/max doesn't contains the filter target, we don't need to search the bucket any more.

### Tests

Add test.

### API and Format

The may speed up select in some conditions. No other influence to user.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 1]
```diff
@@ -275,6 +276,20 @@ private Pair<Snapshot, List<ManifestEntry>> doPlan(
...
             }
         }
+
+        // We group files by bucket here, and filter them by the whole bucket filter.
+        // Why do this: because in primary key table, we can't just filter the value
+        // by the stat in files (see `ChangelogWithKeyFileStoreTable.nonPartitionFilterConsumer`),
+        // but we can do this by filter the whole bucket files
+        files =
+                files.stream()
+                        .collect(Collectors.groupingBy(ManifestEntry::bucket))
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:286]: Partition -> bucket

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 2]
```diff
@@ -339,6 +354,9 @@ private boolean filterByLevel(ManifestEntry entry) {
     /** Note: Keep this thread-safe. */
     protected abstract boolean filterByStats(ManifestEntry entry);
 
+    /** Note: Keep this thread-safe. */
+    protected abstract boolean filterByStats(List<ManifestEntry> entries);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:358]: `filterWholeBucketByStats`

**Comment by JingsongLi** [comment]
Thanks @leaves12138 , left comments above.

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java (hunk 2 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java]: done

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java (hunk 1 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java]: done, added test for this

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**